core/unit_test/TestRangePolicyConstructors.hpp

```diff
@@ -121,4 +122,42 @@ TEST(TEST_CATEGORY_DEATH, range_policy_invalid_bounds) {
 #endif
 }
 
+TEST(TEST_CATEGORY_DEATH, range_policy_implicitly_converted_bounds) {
+  using UIntIndexType = Kokkos::IndexType<unsigned>;
+  using IntIndexType  = Kokkos::IndexType<int>;
+  using UIntPolicy    = Kokkos::RangePolicy<TEST_EXECSPACE, UIntIndexType>;
+  using IntPolicy     = Kokkos::RangePolicy<TEST_EXECSPACE, IntIndexType>;
+
+  std::string msg =
+      "Kokkos::RangePolicy bound type error: unsafe implicit conversion may "
+      "not preserve the original value.\n";
+
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  // signed to unsigned
+  ASSERT_DEATH({ (void)UIntPolicy(-1, 10); }, msg);
+  ASSERT_DEATH({ (void)IntPolicy(0u, std::numeric_limits<unsigned>::max()); },
+               msg);
+  ASSERT_DEATH({ (void)IntPolicy(0LL, std::numeric_limits<long long>::max()); },
+               msg);
+  ASSERT_DEATH({ (void)UIntPolicy(-1, 10, Kokkos::ChunkSize(2)); }, msg);
+
+#else
+  if (!Kokkos::show_warnings()) {
+    GTEST_SKIP() << "Kokkos warning messages are disabled";
+  }
+
+  {
+    UIntPolicy policy(-1, 10);
+    ASSERT_EQ(policy.begin(), 0u);
+    ASSERT_EQ(policy.end(), 0u);
```

Comment:
Check that warnings are raised if deprecation warnings are enabled and show_warnings() returns true

Reply:
There is a minor problem with this test. When `KOKKOS_ENABLE_DEPRECATED_CODE_4` is not enabled, the `RangePolicy` constructors are not templated on the bound types. So, the bound args are implicitly converted to `member_type` and will not be producing the unsafe conversion warning from this PR. Instead, it will only raise the warning from `check_bounds_validity` check.

In the interest of outputting the warning message even when `KOKKOS_ENABLE_DEPRECATED_CODE_4` is defined, the tentative solution is to use this condition instead:
```
#if !defined(KOKKOS_ENABLE_DEPRECATED_CODE_4) || defined(KOKKOS_ENABLE_DEPRECATION_WARNINGS)
```
See [c8befbd](https://github.com/kokkos/kokkos/pull/6754/commits/c8befbdfa8c752ac885edfa5a1f3c4379c9c7654). I don't think this is a desired flag condition to leave it as.